clients/src/main/java/org/apache/kafka/common/protocol/types/Schema.java

```diff
@@ -63,15 +63,32 @@ public void write(ByteBuffer buffer, Object o) {
         }
     }
 
-    /**
-     * Read a struct from the buffer
-     */
     @Override
     public Struct read(ByteBuffer buffer) {
+        return read(buffer, false);
+    }
+
+    /**
+     * Read a struct from the buffer. Missing fields at the end of the buffer are replaced with
+     * their default values if such fields are optional; otherwise a {@code SchemaException} is
+     * thrown to signify that mandatory fields are missing.
+     */
+    public Struct read(ByteBuffer buffer, boolean tolerateMissingWithDefaults) {
         Object[] objects = new Object[fields.length];
         for (int i = 0; i < fields.length; i++) {
             try {
-                objects[i] = fields[i].def.type.read(buffer);
+                if (tolerateMissingWithDefaults) {
```

Comment:
Would it make sense to push this down?
```
                    } else if (tolerateMissingWithDefaults && fields[i].def.hasDefaultValue) {
                        objects[i] = fields[i].def.defaultValue;
                    } else {
```

Reply:
Not sure I follow the partial snippet. 
if `tolerateMissingWithDefaults` is `false` I reinstated the previous state of the code. Which will throw `BufferUnderflowException` if there are missing fields, and will be wrapped in a `SchemaException`. 

Only if `tolerateMissingWithDefaults` is `true` I check whether the buffer has remaining bytes, and if not try to assign a default if it exists, otherwise I throw a `SchemaException` that is again wrapped immediately in another `SchemaException`. 

Would you prefer to have the 
`if (!tolerateMissingWithDefaults)` first?

Reply:
To clarify, what I was suggesting is this:
```
                    if (buffer.hasRemaining()) {
                        objects[i] = fields[i].def.type.read(buffer);
                    } else if (tolerateMissingWithDefaults && fields[i].def.hasDefaultValue) {
                        objects[i] = fields[i].def.defaultValue;
                    } else {
                        throw new SchemaException("Missing value for field '" + fields[i].def.name +
                                "' which has no default value");
                    }
```
I was just trying to consolidate the logic, but it's not a big deal.

By the way, since you drew my attention to it, it might be worth catching the `SchemaException` that we throw here and and raising directly rather than wrapping it in another SchemaException.

Reply:
Ok. I thought we wanted to keep the behavior of throwing `BufferUnderflowException`. But you are right. It's wrapped anyways.

Reply:
Regarding keeping the original exception and throwing that instead of re-wrapping by discarding the stacktrace, I feel that wrapping was intentional because `SchemaException` might be thrown by `Type#read` as well. But then it might not have a message as the catch block indicates here. 

Should I keep the wrapping  but also add the inner exception as a cause? (Will have to add a `SchemaException(String message, Throwable cause)` constructor for that)